src/atomic_densities.jl

```diff
@@ -0,0 +1,241 @@
+              "been specified in any of the given elements / atoms. Your SCF will likely " *
+              "not converge to a spin-broken solution.")
+        return zeros(T, basis.fft_size)
+    end
+
+    @assert length(magmoms) == length(basis.model.atoms)
+    coefficients = map(zip(basis.model.atoms, magmoms)) do (atom, magmom)
+        iszero(magmom[1:2]) || error("Non-collinear magnetization not yet implemented")
+        magmom[3] ≤ n_elec_valence(atom) || error(
+            "Magnetic moment $(magmom[3]) too large for element $(atomic_symbol(atom)) " *
+            "with only $(n_elec_valence(atom)) valence electrons."
+        )
+        magmom[3] / n_elec_valence(atom)
+    end
+
+    form_factors = _valence_density_form_factors(basis, method)
+    atomic_density_superposition(basis, form_factors; coefficients)
+end
+
+function _guess_total_density(basis::PlaneWaveBasis{T}, method::GuessDensityMethod;
+                              coefficients=ones(T, length(basis.model.atoms))
+                             )::Array{T,3} where {T}
+    form_factors = _valence_density_form_factors(basis, method)
+    atomic_density_superposition(basis, form_factors; coefficients)
+end
+
+function atomic_density_superposition(basis::PlaneWaveBasis{T},
+                                       form_factors::IdDict{Tuple{Int,T},T};
+                                       coefficients=ones(T, length(basis.model.atoms))
+                                       )::Array{T,3} where {T}
+    model = basis.model
+    G_cart = G_vectors_cart(basis)
+
+    ρ = map(enumerate(G_vectors(basis))) do (iG, G)
+        Gnorm = norm(G_cart[iG])
+        ρ_iG = sum(enumerate(model.atom_groups); init=zero(Complex{T})) do (igroup, group)
+            sum(group) do iatom
+                structure_factor::Complex{T} = cis2pi(-dot(G, model.positions[iatom]))
+                coefficients[iatom]::T * form_factors[(igroup, Gnorm)]::T * structure_factor
+            end
+        end
+        ρ_iG / sqrt(model.unit_cell_volume)
+    end
+    enforce_real!(basis, ρ)  # Symmetrize Fourier coeffs to have real iFFT
+    irfft(basis, ρ)
+end
+
+function _valence_density_form_factors(basis::PlaneWaveBasis{T}, ::GaussianGuessDensity,
+                                      )::IdDict{Tuple{Int,T},T} where {T<:Real}
+    model = basis.model
+    form_factors = IdDict{Tuple{Int,T},T}()  # IdDict for Dual compatability
+    for G in G_vectors_cart(basis)
+        Gnorm = norm(G)
+        for (igroup, group) in enumerate(model.atom_groups)
+            if !haskey(form_factors, (igroup, Gnorm))
+                element = model.atoms[first(group)]
+                form_factor = gaussian_valence_charge_density_fourier(element, Gnorm)
+                form_factors[(igroup, Gnorm)] = form_factor
+            end
+        end
+    end
+    form_factors
+end
+
+function _valence_density_form_factors(basis::PlaneWaveBasis{T}, ::PspGuessDensity,
```

Comment:
There is a lot of dupicate code here that could be solved by multiple dispatch, see my ideas below.